mtags/src/main/scala/scala/meta/internal/pc/TypeDefinitionProvider.scala

```diff
@@ -0,0 +1,152 @@
+package scala.meta.internal.pc
+
+import org.eclipse.{lsp4j => l}
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.pc.OffsetParams
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.internal.metals.CompilerOffsetParams
+
+class TypeDefinitionProvider(val compiler: MetalsGlobal, params: OffsetParams) {
+  import compiler._
+
+  def typedTree: Option[(RichCompilationUnit, Position, Tree)] = {
+    if (params.isWhitespace) {
+      None
+    } else {
+      val unit = addCompilationUnit(
+        code = params.text(),
+        filename = params.filename(),
+        cursor = None
+      )
+      val pos = unit.position(params.offset())
+      val tree = typedHoverTreeAt(pos)
+      Some(unit, pos, tree)
+    }
+  }
+
+  def typeSymbol: Option[(RichCompilationUnit, Position, Symbol)] = {
+    val tri = typedTree
+    tri match {
+      case Some((t1, t2, tree))
+          if tree.symbol.isTypeSymbol || tree.symbol.isMethod || tree.symbol.isConstructor =>
+        Some(t1, t2, tree.symbol)
+      case Some((t1, t2, tree)) if tree.tpe.isDefined =>
```

Comment:
I tihnk using simple `if/else` might be simpler here. We can `flatMap` over typedTree itself.